2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java

```diff
@@ -8,19 +8,24 @@
 class NeoPixelsIO extends I2C {
 	
 	private final Packet displayPacket;
+	private final ByteBuffer buffer;
+	
+	private final int MAX_PACKET_SIZE = 16;
 	
 	public NeoPixelsIO(Port port, int deviceAddress) {
 		super(port, deviceAddress);
 		displayPacket = new Packet();
+		buffer = ByteBuffer.allocateDirect(MAX_PACKET_SIZE);
 	}
 	
 	public boolean writePacket(Packet packet) {
-		return writeBulk(packet.fillBuffer(), packet.getPacketSize());
+		packet.fillBuffer(buffer);
+		return writeBulk(buffer, packet.getPacketSize());
 	}
 	
 	public boolean sendSyncPacket() {
-		ByteBuffer packet = Packet.syncPacket();
-		return writeBulk(packet, packet.capacity());
+		Packet.syncPacket(buffer);
+		return writeBulk(buffer, buffer.capacity());
```

Comment:
Why is the sending of the SyncPacket any different than sending a ShowPacket, or any other packet?  The IO class should only know about generic Packet and how to bundle it into a ByteBuffer and write it to I2C.

I understand the desire to have pre-made Sync and Show packets lying around, but they should be owned outside the IO class.  Perhaps in the NeoPixels class.  But they should be sent via the writePacket method just like any other packet.

Reply:
If we use a Factory to make Packets, then the only function of the Packet is to hold a series of bytes, and to calculate the size of that series (so I might just use an array and delete Packet entirely). Then the writePacket method just puts the size as the first byte in the ByteBuffer and fills the rest of the ByteBuffer with the data held by the Packet. Since the maximum size of a packet (including the wrapper) is 16 bytes, that's what I'm using as the size of my ByteBuffer.
However, the structure of a Sync packet is different from that of all other packets. It doesn't have the size byte, it's just 16 (or more) bytes of 0xFF. So I want to fill my entire ByteBuffer with 0xFF. The three approaches I can think of are:
1. Have writePacket method detect when the packet is a Sync packet and not write the size byte
1. Put a separate method on NeoPixelsIO that to send a Sync packet that just fills the ByteBuffer with 0xFF
1. Make the ByteBuffer 17 bytes long and have the packet data be 16 bytes of 0xFF.

I elected to go with the second option, but if you have a better idea I'm all ears

Reply:
Now that I think about it, I could have the Factory just return byte arrays for the NeoPixelsIO to load into the buffer, and remove Packet entirely. Then, the factory would be in charge of setting the size byte, and it could have a method to create a synchronization packet that would just be a byte array of length 16 filled with 0xFF.